test/e2e/ocpbugs48050_test.go

```diff
@@ -0,0 +1,611 @@
+		return nil, fmt.Errorf("failed to create Kubernetes client: %v", err)
+	}
+
+	routeClient, err := routev1client.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create route client: %v", err)
+	}
+
+	prometheusClient, err := metrics.NewPrometheusClient(context.TODO(), kubeClient, routeClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Prometheus client: %v", err)
+	}
+
+	return prometheusClient, nil
+}
+
+func queryPrometheus(client prometheusv1client.API, query string, timeout time.Duration) (model.Value, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	result, warnings, err := client.Query(ctx, query, time.Now())
+	if err != nil {
+		return nil, fmt.Errorf("error querying Prometheus: %v", err)
+	}
+
+	if len(warnings) > 0 {
+		return nil, fmt.Errorf("warnings querying Prometheus: %v", warnings)
+	}
+
+	return result, nil
+}
+
+type podScrapeCounts map[string]float64
+
+func getTotalScrapesForPods(promClient prometheusv1client.API, timeout time.Duration) (podScrapeCounts, error) {
+	query := `sum by (pod) (haproxy_exporter_total_scrapes{pod=~"router-default-.*"})`
+
+	result, err := queryPrometheus(promClient, query, timeout)
+	if err != nil {
+		return nil, err
+		//t.Fatalf("Failed to query Prometheus for total scrapes: %v", err)
+	}
+
+	vector, ok := result.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("expected result type model.Vector, got %T", result)
+	}
+
+	scrapeCounts := make(map[string]float64)
+
+	for _, sample := range vector {
+		podName := string(sample.Metric["pod"])
+		scrapeCounts[podName] = float64(sample.Value)
+	}
+
+	return scrapeCounts, nil
+}
+
+func waitForDefaultRouterScrapesIncrement(promClient prometheusv1client.API, minNewScrapes int, timeout time.Duration, progress func(initialCounts, currentCounts podScrapeCounts, allIncremented bool)) error {
+	initialCounts, err := getTotalScrapesForPods(promClient, timeout)
+	if err != nil {
+		return fmt.Errorf("failed to get initial scrape counts: %v", err)
+	}
+
+	deadline := time.Now().Add(timeout)
+
+	for time.Now().Before(deadline) {
+		currentCounts, err := getTotalScrapesForPods(promClient, timeout)
+		if err != nil {
+			return fmt.Errorf("failed to get current scrape counts: %v", err)
+		}
+
+		allIncremented := true
+		for podName, initialCount := range initialCounts {
+			if currentCount, exists := currentCounts[podName]; !exists || currentCount < initialCount+float64(minNewScrapes) {
+				allIncremented = false
+				break
+			}
+		}
+
+		// Call the progress function with the current state
+		if progress != nil {
+			progress(initialCounts, currentCounts, allIncremented)
+		}
+
+		if allIncremented {
+			return nil
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+
+	return fmt.Errorf("timeout after %v waiting for all pods to increment total scrapes by at least %d", timeout, minNewScrapes)
+}
+
+func getCanaryImageFromIngressOperatorDeployment() (string, error) {
```

Comment:
No - or at least I don't care. This is EOL code.

Reply:
> I also want a smooth backport to 4.12, with no complications or unnecessary changes, ensuring a clean merge. And for that reason, I'd rather leave everything in new files (where practical).

I see, makes sense.

Reply:
Keep in mind that any future backports may start from 4.18. Ideally, we want to avoid merge conflicts when going all the way back to 4.13 (and 4.12), and keeping these changes isolated may well help us. This specific change is starting out at 4.13. By isolating changes, we reduce the risk of conflicts, especially when merging across versions that share fewer common ancestors in git.